peps/pep-0705.rst

```diff
@@ -310,235 +217,260 @@ This is consistent with the behavior introduced in :pep:`655`.
-
-* A TypedDict type ``A`` with no key ``'x'`` that does not allow other keys may be consistent with a TypedDict type with a read-only, non-required key ``'x'``. Example::
-
-    class A(TypedDict, total=False, readonly=True):
-        y: int
+* A TypedDict type ``A`` with no explicit key ``'x'`` is not consistent with a TypedDict type ``B`` with a non-required key ``'x'``, since at runtime the key ``'x'`` could be present and have an incompatible type (which may not be visible through ``A`` due to structural subtyping). The only exception to this rule is if the item in ``B`` is read-only, and the value type is a top type (``object``, ``Any`` or pylance's ``Unknown``). For example::
 
-    class B(TypedDict, other_keys=Never):
+    class A(TypedDict):
         x: int
 
-    def f(a: A) -> int:
-       return a.get("y", 0)
+    class B(TypedDict):
+        x: int
+        y: ReadOnly[NotRequired[object]]
 
-    def g(b: B) -> None:
-       b["x"] = f(b)  # Accepted by type checker
+    a: A = { "x": 1 }
+    b: B = a  # Accepted by type checker
 
-Union Operation
+Merge operation
 ---------------
 
-The union operation creates a new dictionary with the merged keys and values of its two operands. As such, the result should be consistent with any type that can hold the possible key-value pairs, not just types compatible with the operand types. For example::
+The merge operation (``d1 | d2``) creates a new dictionary with the merged keys and values of its two operands. As such, the result should be consistent with any type that can hold the possible items. For example::
 
-    class A(TypedDict, readonly=True, other_keys=Never):
-        x: int
+    class A(TypedDict):
+        x: ReadOnly[int]
 
-    class B(TypedDict, total=False, readonly=True, other_keys=Never):
-        x: str
+    class B(TypedDict):
+        x: ReadOnly[NotRequired[str]]
 
     class C(TypedDict):
         x: int | str
 
     def union_a_b(a: A, b: B) -> C:
-        # Accepted by type-checker, even though C is not read-only and
-        # allows other keys:
+        # Accepted by type-checker, even though A, B and C are mutually inconsistent
         return a | b
 
-This is different from the usual compatibility rules, where the result of an operation has a defined type which the variable it is assigned to must be consistent with. A similar situation occurs with ``TypedDict`` and ``copy()`` or ``deepcopy()``.
+Type checkers should conform to the following rules, even if they are not expressed correctly in the typeshed due to language constraints.
 
-If the union of two TypedDict objects of type ``A`` and ``B`` are assigned to a TypedDict of type ``C``, the type checker should verify that:
+If the merger of two TypedDict objects of type ``A`` and ``B`` are assigned to a TypedDict of type ``C``, the type checker should verify that:
```

Comment:
My contention is that is incorrect for structural types. Any type I could try to answer your question with would rule out valid assignments.

Reply:
In general, if you cannot describe the type that results from an operation, then a type checker cannot determine whether that type is compatible with another type.

I think what you're saying is that you want the spec to define rules for [bidirectional type inference](https://microsoft.github.io/pyright/#/type-inference?id=bidirectional-type-inference-expected-types). This applies only when there is an "expected type" (or what mypy calls a "type context").

The challenge here is that bidirectional type inference isn't currently part of the typing spec, and type checkers are not required to support it. I think all of the major type checkers implement some form of it. Your proposed language would require a type checker to support this if it wants to be compliant with the spec. This would be the first time that any typing spec introduces this concept — and mandates that it's used. I'd be interested to get input from other type checker maintainers.

Even if the spec does include rules for bidirectional type inference, that still begs the question about what behavior is mandated (or recommended) in cases where there is no "expected type", as in my code sample above. Should `a | b` produce an error in this case? If not, what type should `a | b` evaluate to? Perhaps the answer is `Any` for now and `A & B` once intersections are supported?

Reply:
For instance, given the following:

```python
class A(TypedDict):
  foo: int

class B(TypedDict):
  foo: SOME_TYPE

a: A = { "foo" : 1 }
b: B = copy(a)  # or a.copy(), or (a | a), or deepcopy(a)
```

The following are all completely valid declarations that could be used for `SOME_TYPE`:
* `int`
* `ReadOnly[int]`
* `int | None`
* `NotRequired[int]`
* `ReadOnly[int | str | B]`

There's no type I could give that's structurally compatible with all those.

Reply:
One option is to not mandate this behavior. If you leave it optional in the spec (i.e. change "should" to "may"), it would probably alleviate any concerns from type checker maintainers.

Reply:
> Even if the spec does include rules for bidirectional type inference, that still begs the question about what behavior is mandated (or recommended) in cases where there is no "expected type", as in my code sample above. Should a | b produce an error in this case? If not, what type should a | b evaluate to? Perhaps the answer is Any for now and A & B once intersections are supported?

I don't have a good theoretical background for this, but I want to say "assume it could be anything the assignment rules allow, and work it out from what happens below". I think pyright does this in other places?

> One option is to not mandate this behavior. If you leave it optional in the spec (i.e. change "should" to "may"), it would probably alleviate any concerns from type checker maintainers.

Absolutely happy to do that. Do you want me to do it now, or after we get input from other type checker maintainers? If we don't do it now, I can still add an explanation to the "rejected alternatives" section so it's clear we've thought about it and are looking for feedback.

Reply:
> I don't have a good theoretical background for this, but I want to say "assume it could be anything the assignment rules allow, and work it out from what happens below". I think pyright does this in other places?

The assignment rules allow anything when the target of the assignment has no declared type. For example, `x = a | b` where `x` has no type annotation. In this case, we need to decide 1) is the operation `a | b` allowed, or should it produce a type error, and 2) if it is allowed, what is type does `a | b` evaluate to?

Reply:
It would be really nice if I could do something like:

```python
x = a | b
x["foo"] = 3
f(x)
g(x)
```

and have the type system figure out whether this is legal without needing to first decide on a "type" for x. Maybe "bar" is of type `int` in a, `str` in b, and mutable `int | str | None` on f -- that's ok! We can "work backwards" and decide the return type of `a | b` has a mutable "bar" of value type `int | str | None`.

If this is a bad idea (it probably is!), then perhaps we could construct a fallback type `F` that hopefully allows common idiomatic code:

* Every key in A or B is in F
* If any of these keys is missing in B, or not required in B and missing in A, error
* A key is required in F if it is required in either A or B
* The associated value type in F is the same as in B if it is required in B
* Otherwise it is the union of the value types in A and B
* Nothing is read-only

I think this is likely to be safer than `Any`, and matches what I believe is existing behaviour for a non-read-only TypedDict `A` when you do `A | A` (you get back `A`). In the cases where it's wrong, you can still create a type by hand.

Reply:
I'm also uneasy about mandating that type checkers infer a type that cannot be expressed in the normal type system.

Reply:
Eric seems happy to bring this as-is to the wider audience to get opinions on whether mandating bidirectional type inference is ok. I will add it explicitly to a new "Open Issues" section as well before we merge. Are you ok with that @JelleZijlstra?

Reply:
This isn't the most important part of this proposal to me, so I'm also ok with changing it to be non-mandatory, especially if this would be a blocker to mypy support.

Reply:
I went ahead and drafted this change up here: https://github.com/alicederyn/peps/pull/8

Reply:
Thanks for the feedback on the draft! I've removed the sections about merge and copy/deepcopy and added my reasoning to the Rejected Alternatives section.